pkg/providers/amifamily/bootstrap/nodeadm.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/json"
+	"sigs.k8s.io/yaml"
+
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily/bootstrap/mime"
+)
+
+type Nodeadm struct {
+	Options
+}
+
+func (n Nodeadm) Script() (string, error) {
+	nodeConfigYAML, err := n.getNodeConfigYAML()
+	if err != nil {
+		return "", fmt.Errorf("generating NodeConfig, %w", err)
+	}
+	customEntries, err := n.parseUserData()
+	if err != nil {
+		return "", fmt.Errorf("parsing custom UserData, %w", err)
+	}
+	mimeArchive := mime.Archive(append([]mime.Entry{{
+		ContentType: mime.ContentTypeNodeConfig,
+		Content:     nodeConfigYAML,
+	}}, customEntries...))
+	userData, err := mimeArchive.Serialize()
+	if err != nil {
+		return "", err
+	}
+	return userData, nil
+}
+
+// getNodeConfigYAML returns the Karpenter generated NodeConfig YAML object serialized as a string
+func (n Nodeadm) getNodeConfigYAML() (string, error) {
+	config := &admv1alpha1.NodeConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       admapi.KindNodeConfig,
+			APIVersion: admv1alpha1.GroupVersion.String(),
+		},
+		Spec: admv1alpha1.NodeConfigSpec{
+			Cluster: admv1alpha1.ClusterDetails{
+				Name:              n.ClusterName,
+				APIServerEndpoint: n.ClusterEndpoint,
+				CIDR:              n.ClusterCIDR,
+			},
+		},
+	}
+	if lo.FromPtr(n.CABundle) != "" {
+		ca, err := base64.StdEncoding.DecodeString(*n.CABundle)
+		if err != nil {
+			return "", fmt.Errorf("decoding CABundle, %w", err)
+		}
+		config.Spec.Cluster.CertificateAuthority = ca
+	}
+	inlineConfig, err := n.generateInlineKubeletConfiguration()
+	if err != nil {
+		return "", err
+	}
+	config.Spec.Kubelet.Config = inlineConfig
+	config.Spec.Kubelet.Flags = n.getExtraKubeletFlags()
+
+	// Convert to YAML at the end for improved legibility.
+	configYAML, err := yaml.Marshal(config)
+	if err != nil {
+		return "", err
+	}
+	return string(configYAML), nil
+}
+
+// generateInlineKubeletConfiguration returns a serialized form of the KubeletConfiguration specified by the Nodeadm
+// options, for use with nodeadm's NodeConfig struct.
+func (n Nodeadm) generateInlineKubeletConfiguration() (map[string]runtime.RawExtension, error) {
+	kubeConfigJSON, err := json.Marshal(n.KubeletConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeConfigMap := map[string]runtime.RawExtension{}
+	err = json.Unmarshal(kubeConfigJSON, &kubeConfigMap)
+	if err != nil {
+		return nil, err
+	}
+	if len(n.Taints) != 0 {
+		kubeConfigMap["registerWithTaints"] = runtime.RawExtension{
+			Raw: lo.Must(json.Marshal(n.Taints)),
+		}
+	}
+	return kubeConfigMap, nil
+}
+
+// getExtraKubeletFlags returns the kubelet flags derived from the Nodeadm options that could not be represented in the
+// kubelet config.
+func (n Nodeadm) getExtraKubeletFlags() []string {
+	args := []string{}
+	if labelArg := n.nodeLabelArg(); labelArg != "" {
```

Comment:
I could go either way, but since this list is so small now do you think that there is a need for a separate function?

Reply:
I'm mixed as well, was considering inlining it when I inlined the kubeletConfigurationExtraFields. It keeps the calling function a little cleaner but given that this list should never expand there's not much benefit in factoring out. Probably fine to just go ahead and inline.

Reply:
I'm good either way. Feel free to resolve this comment based on what you decide.